prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/FactoryParserHandler.java

```diff
@@ -0,0 +1,709 @@
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles the common code of device factory
+ */
+@Component(immediate = true, service = IDeviceMappingHandler.class)
+public class FactoryParserHandler implements IDeviceMappingHandler {
+
+    /**
+     * Logger
+     */
+    private static final Logger logger = LoggerFactory.getLogger(FactoryParserHandler.class);
+
+    /**
+     * Keeps track of the state of the current record
+     */
+    private static class RecordState {
+        /**
+         * Place holder -&gt; record path
+         */
+        Map<String, RecordPath> placeholders;
+
+        /**
+         * Variable name -&gt; unresolved record path
+         */
+        Map<String, RecordPath> rawVariables;
+
+        /**
+         * Variable name -&gt; resolved record path
+         */
+        Map<String, String> variables;
+
+        /**
+         * SensiNact resource path -&gt; record path
+         */
+        List<ResourceMapping> rcMappings;
+    }
+
+    /**
+     * SensiNact update endpoint
+     */
+    @Reference
+    private PrototypePush prototypePush;
+
+    /**
+     * Bundle context
+     */
+    private BundleContext context;
+
+    /**
+     * Component activated
+     */
+    @Activate
+    void activate(ComponentContext context) {
+        this.context = context.getBundleContext();
+    }
+
+    /**
+     * Component deactivated
+     */
+    @Deactivate
+    void deactivate() {
+        this.context = null;
+    }
+
+    /**
+     * Looks for the service reference of the parser with the given ID
+     */
+    private ServiceReference<IDeviceMappingParser> findParser(final String parserId) throws MissingParserException {
+        final Collection<ServiceReference<IDeviceMappingParser>> svcRefs;
+        try {
+            svcRefs = context.getServiceReferences(IDeviceMappingParser.class,
+                    String.format("(%s=%s)", IDeviceMappingParser.PARSER_ID, parserId));
+        } catch (InvalidSyntaxException e) {
+            throw new MissingParserException(String.format("Invalid parser ID '%s': %s'", parserId, e.getMessage()));
+        }
+
+        if (svcRefs == null || svcRefs.isEmpty()) {
+            throw new MissingParserException(String.format("Parser ID '%s' is missing", parserId));
+        }
+
+        return svcRefs.stream().sorted(Comparator.naturalOrder()).findFirst().get();
+    }
+
+    @Override
+    public void handle(final SensiNactSession session, final DeviceMappingConfigurationDTO configuration,
```

Comment:
I think that the `SensiNactSession` can be removed from here now? If not then what user does it act as?

Reply:
Agreed